cmd/juju/machine/add.go

```diff
@@ -168,7 +169,47 @@ type MachineManagerAPI interface {
 	Close() error
 }
 
-var manualProvisioner = manual.ProvisionMachine
+// manualProvisioner provisions a machine agent to an existing host, via
+// an SSH connection or WinRM connection to the specified host. The host may optionally be preceded
+// with a login username, as in [user@]host.
+//
+// On successful completion, this function will return the id of the state.Machine
+// that was entered into state.
+//
+// we use manualProvisioner as a lambda and not a function
+// because this variable will be overwritten in tests.
+var manualProvisioner = func(args manual.ProvisionMachineArgs,
+	placement *instance.Placement) (machineId string, err error) {
+	defer func() {
+		if machineId != "" && err != nil {
```

Comment:
If the provisioning function returns an error, we should disregard any other results. The provisioning function needs to be responsible for cleaning up after itself.